## T12B model: Neutral matrix

In [1]:
import numpy as np
import pandas as pd
import subprocess
import time

%matplotlib inline
import matplotlib.pyplot as plt 

In [2]:
#good plots
plt.rcParams.update({'font.size': 18}) 
#plt.rc('font',size='18')

In [6]:
def MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2):    
    
    dcOut={}
    
    #Diagonalization of Mchi matrix by the bi-unitary transfortion V and U
    
    MX0 = np.matrix( [[MN,0.,1./np.sqrt(2.)*vevSM*y2],
                      [0.,MPsi,l1],
                      [1./np.sqrt(2.)*vevSM*y1,l2,Meta]])
    
    #squared eigenvalues e eigenvectors for the V MATRIX
    (MVdiag2,V)=np.linalg.eig(MX0*np.transpose(MX0))
    #squared eigenvalues e eigenvectors for the U MATRIX
    (MUdiag2,U)=np.linalg.eig(np.transpose(MX0)*MX0)


    M1=np.sqrt(np.abs(MVdiag2[0]))
    M2=np.sqrt(np.abs(MVdiag2[1]))
    M3=np.sqrt(np.abs(MVdiag2[2]))

    V11=V[0,0]
    V12=V[0,1]
    V13=V[0,2]
    V21=V[1,0]
    V22=V[1,1]
    V23=V[1,2]
    V31=V[2,0]
    V32=V[2,1]
    V33=V[2,2] 
    U11=U[0,0]
    U12=U[0,1]
    U13=U[0,2]
    U21=U[1,0]
    U22=U[1,1]
    U23=U[1,2]
    U31=U[2,0]
    U32=U[2,1]
    U33=U[2,2]   

    dcOut['M1']= M1
    dcOut['M2']= M2  
    dcOut['M3']= M3  
    dcOut['V11']= V11 
    dcOut['V12']= V12 
    dcOut['V13']= V13 
    dcOut['V21']= V21 
    dcOut['V22']= V22 
    dcOut['V23']= V23 
    dcOut['V31']= V31 
    dcOut['V32']= V32 
    dcOut['V33']= V33 
    dcOut['U11']= U11 
    dcOut['U12']= U12 
    dcOut['U13']= U13 
    dcOut['U21']= U21 
    dcOut['U22']= U22 
    dcOut['U23']= U23 
    dcOut['U31']= U31 
    dcOut['U32']= U32 
    dcOut['U33']= U33 
            
    return dcOut

In [7]:
Gf = 1.16637000E-05
vevSM = 1./np.sqrt(np.sqrt(2.)*Gf)

In [8]:
#benchmarck point
MN = 101.0
MPsi = 200.1
Meta = 301.1
l1 = 1.02e-2
l2 = 2.01e-3
y1 = 1.01e-2
y2 = 2.1e-3

MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2)

{'M1': 100.9955695320795,
 'M2': 301.10684318118194,
 'M3': 200.0999996644955,
 'U11': 0.99997522445007936,
 'U12': -0.0070392105558769692,
 'U13': 8.7372602477659586e-07,
 'U21': 5.0579835385128309e-07,
 'U22': -5.2270094528977737e-05,
 'U23': -0.99999999863379085,
 'U31': -0.0070392105919296812,
 'U32': -0.99997522308434583,
 'U33': 5.2265239087673537e-05,
 'V11': 0.99999360832362982,
 'V12': -0.0035753757283830279,
 'V13': 5.3650918272463957e-07,
 'V21': 2.9120388848907209e-07,
 'V22': -6.8610174954878976e-05,
 'V23': -0.99999999764627967,
 'V31': -0.0035753757567775807,
 'V32': -0.99999360597008058,
 'V33': 6.8608695257419038e-05}

'''
    56     1.00995570E+02  # Fn_1
    57     2.00100000E+02  # Fn_2
    58     3.01106843E+02  # Fn_3
        
Block LNEUTROMIX Q=  1.73500000E+02  # ()
  1  1     9.99993608E-01   # Real(RL(1,1),dp) V11': 0.99999360832362982,
  1  2     2.91203888E-07   # Real(RL(1,2),dp) V21': 2.9120388848907209e-07,
  1  3    -3.57537576E-03   # Real(RL(1,3),dp) V31': -0.0035753757567775807,
  2  1    -5.36509183E-07   # Real(RL(2,1),dp) V13': 5.3650918272463957e-07
  2  2     9.99999998E-01   # Real(RL(2,2),dp) V23': -0.99999999764627967
  2  3    -6.86086953E-05   # Real(RL(2,3),dp) V33': 6.8608695257419038e-05
  3  1     3.57537573E-03   # Real(RL(3,1),dp) V12': -0.0035753757283830279,
  3  2     6.86101750E-05   # Real(RL(3,2),dp) V22': -6.8610174954878976e-05
  3  3     9.99993606E-01   # Real(RL(3,3),dp) V32': -0.99999360597008058         
Block RNEUTROMIX Q=  1.73500000E+02  # ()
  1  1     9.99975224E-01   # Real(RR(1,1),dp)
  1  2     5.05798354E-07   # Real(RR(1,2),dp)
  1  3    -7.03921059E-03   # Real(RR(1,3),dp)
  2  1    -8.73726025E-07   # Real(RR(2,1),dp)
  2  2     9.99999999E-01   # Real(RR(2,2),dp)
  2  3    -5.22652391E-05   # Real(RR(2,3),dp)
  3  1     7.03921056E-03   # Real(RR(3,1),dp)
  3  2     5.22700945E-05   # Real(RR(3,2),dp)
  3  3     9.99975223E-01   # Real(RR(3,3),dp)
'''

In [9]:
def MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2):    
    
    dcOut={}
    
    #Diagonalization of Mchi matrix by the bi-unitary transfortion V and U
    
    MX0 = np.matrix( [[MN,0.,1./np.sqrt(2.)*vevSM*y2],
                      [0.,MPsi,l1],
                      [1./np.sqrt(2.)*vevSM*y1,l2,Meta]])
    
    #squared eigenvalues e eigenvectors for the V MATRIX
    (MVdiag2,V)=np.linalg.eig(MX0*np.transpose(MX0))
    #squared eigenvalues e eigenvectors for the U MATRIX
    (MUdiag2,U)=np.linalg.eig(np.transpose(MX0)*MX0)


    M1=np.sqrt(np.abs(MVdiag2[0]))
    M2=np.sqrt(np.abs(MVdiag2[1]))
    M3=np.sqrt(np.abs(MVdiag2[2]))  
    
    #Order in the eigenstates and eigenvalues
    if M1 < M2 and M1 < M3:
        m1 = M1
        
        V11=V[0,0]
        V12=V[1,0]
        V13=V[2,0]
        U11=U[0,0]
        U12=U[1,0]
        U13=U[2,0]

        if M2 < M3:
            m2 = M2
            m3 = M3
        else:
            m2 = M3
            m3 = M2  
            
            V21=V[0,2]
            V22=V[1,2]
            V23=V[2,2]          
            V31=V[0,1]
            V32=V[1,1]
            V33=V[2,1] 
            U21=U[0,2]
            U22=U[1,2]
            U23=U[2,2]          
            U31=U[0,1]
            U32=U[1,1]
            U33=U[2,1] 


    dcOut['M1']= m1
    dcOut['M2']= m2  
    dcOut['M3']= m3  
    dcOut['V11']= V11 
    dcOut['V12']= V12 
    dcOut['V13']= V13 
    dcOut['V21']= V21 
    dcOut['V22']= V22 
    dcOut['V23']= V23 
    dcOut['V31']= V31 
    dcOut['V32']= V32 
    dcOut['V33']= V33 
    dcOut['U11']= U11 
    dcOut['U12']= U12 
    dcOut['U13']= U13 
    dcOut['U21']= U21 
    dcOut['U22']= U22 
    dcOut['U23']= U23 
    dcOut['U31']= U31 
    dcOut['U32']= U32 
    dcOut['U33']= U33 
            
    return dcOut

In [10]:
#benchmarck point
MN = 101.0
MPsi = 200.1
Meta = 301.1
l1 = 1.02e-2
l2 = 2.01e-3
y1 = 1.01e-2
y2 = 2.1e-3

MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2)

{'M1': 100.9955695320795,
 'M2': 200.0999996644955,
 'M3': 301.10684318118194,
 'U11': 0.99997522445007936,
 'U12': 5.0579835385128309e-07,
 'U13': -0.0070392105919296812,
 'U21': 8.7372602477659586e-07,
 'U22': -0.99999999863379085,
 'U23': 5.2265239087673537e-05,
 'U31': -0.0070392105558769692,
 'U32': -5.2270094528977737e-05,
 'U33': -0.99997522308434583,
 'V11': 0.99999360832362982,
 'V12': 2.9120388848907209e-07,
 'V13': -0.0035753757567775807,
 'V21': 5.3650918272463957e-07,
 'V22': -0.99999999764627967,
 'V23': 6.8608695257419038e-05,
 'V31': -0.0035753757283830279,
 'V32': -6.8610174954878976e-05,
 'V33': -0.99999360597008058}

'''
Block RNEUTROMIX Q=  1.73500000E+02  # ()
  1  1     9.99975224E-01   # Real(RR(1,1),dp)
  1  2     5.05798354E-07   # Real(RR(1,2),dp)
  1  3    -7.03921059E-03   # Real(RR(1,3),dp)
  2  1    -8.73726025E-07   # Real(RR(2,1),dp)
  2  2     9.99999999E-01   # Real(RR(2,2),dp)
  2  3    -5.22652391E-05   # Real(RR(2,3),dp)
  3  1     7.03921056E-03   # Real(RR(3,1),dp)
  3  2     5.22700945E-05   # Real(RR(3,2),dp)
  3  3     9.99975223E-01   # Real(RR(3,3),dp)
Block LNEUTROMIX Q=  1.73500000E+02  # ()
  1  1     9.99993608E-01   # Real(RL(1,1),dp)
  1  2     2.91203888E-07   # Real(RL(1,2),dp)
  1  3    -3.57537576E-03   # Real(RL(1,3),dp)
  2  1    -5.36509183E-07   # Real(RL(2,1),dp)
  2  2     9.99999998E-01   # Real(RL(2,2),dp)
  2  3    -6.86086953E-05   # Real(RL(2,3),dp)
  3  1     3.57537573E-03   # Real(RL(3,1),dp)
  3  2     6.86101750E-05   # Real(RL(3,2),dp)
  3  3     9.99993606E-01   # Real(RL(3,3),dp)
'''

# Neutrino masses for a benchmark point not working

In [75]:
#!/usr/bin/env python

import pyslha
import pyT12B_LesHouches_generator
import numpy as np
import pandas as pd
import subprocess
import time
import sys

import NEUTRINO2018
import neutrino_analytic3

# Tiempo inicial
t1=time.time() 

#Open
xdict = pyT12B_LesHouches_generator.buildSLHAinFile()

x=[]
    
L1 = 1.3e-1 #Warning
L6 = 1.300E-02

mu31 = 1.01e6
mu32 = 2.01e6
mu33 = 1.01e8
L71 = 1.01e-2
L72 = 2.01e-2
L73 = 3.01e-4

# Modific the LesHouches         
xdict.blocks['MINPAR'].entries[1]='%.6E    # lambda1Input' %L1     
xdict.blocks['MINPAR'].entries[2]='%.6E    # lambda61Input' %L6 
xdict.blocks['MINPAR'].entries[5]='%.6E    # lambda71Input' %L71   
xdict.blocks['MINPAR'].entries[6]='%.6E    # lambda72Input' %L72   
xdict.blocks['MINPAR'].entries[7]='%.6E    # lambda73Input' %L73       
xdict.blocks['MINPAR'].entries[8]='%.6E    # mu31Input' %mu31
xdict.blocks['MINPAR'].entries[9]='%.6E    # mu32Input' %mu32
xdict.blocks['MINPAR'].entries[10]='%.6E    # mu33Input' %mu33


###  NEUTRINO EXPERIMENTAL VALUES ###################################################
#phases of the PMNS matrix and the R 
phases1 = np.random.uniform(0.,0.0*np.pi,3) # WARNING! They are in zero
delta = phases1[0]
eta1 = phases1[1]
eta2 = phases1[2]
#light neutrino masses (up 3 sigma range) NH (NO)
#mnu1 = 10**((np.log10(2.5e-3)-np.log10(1e-9))*np.random.uniform(0,1)+np.log10(1e-9))*1e-9 
mnu1 = 1.0e-20
mnu2 = np.sqrt(np.random.uniform(7.05e-5,8.14e-5)*1.0e-18+mnu1**2)
mnu3 = np.sqrt(np.random.uniform(2.41e-3,2.60e-3)*1.0e-18+mnu1**2)
print("Experimental masses:")
print("mnu1=",mnu1,"mnu2=",mnu2,"mnu3=",mnu3)
#mixing angles (up 3 sigma range) NH
t12 = np.arcsin(np.sqrt(np.random.uniform(0.273,0.379)))
t23 = np.arcsin(np.sqrt(np.random.uniform(0.445,0.599)))
t13 = np.arcsin(np.sqrt(np.random.uniform(0.0196,0.0241)))
#Building PMNS matrix NH
UM12 = np.array([ [np.cos(t12),np.sin(t12),0.], [-np.sin(t12),np.cos(t12),0.], [0.,0.,1.0] ])
UM13 = np.array([ [np.cos(t13),0.,np.sin(t13)], [0.,1.0,0.], [-np.sin(t13),0.,np.cos(t13)] ])
UM23 = np.array([ [1.0,0.,0.], [0.,np.cos(t23),np.sin(t23)], [0.,-np.sin(t23),np.cos(t23)] ])
Uphases = np.array([ [np.exp(eta1*1j),0.,0.], [0.,np.exp(eta2*1j),0.], [0.,0.,1.0] ])
U=np.dot(UM23,np.dot(UM13,np.dot(UM12,Uphases)))
#Defining the U elementes. readeable
U11 = np.real(U[0,0])
U12 = np.real(U[0,1])
U13 = np.real(U[0,2])
U21 = np.real(U[1,0])
U22 = np.real(U[1,1])
U23 = np.real(U[1,2])
U31 = np.real(U[2,0])
U32 = np.real(U[2,1])
U33 = np.real(U[2,2])

### Algoritm to compute LAMBDAi elements
MS1 = np.sqrt(mu31 + 0.5*L71*vevSM**2)
MS2 = np.sqrt(mu32 + 0.5*L72*vevSM**2)
MS3 = np.sqrt(mu33 + 0.5*L73*vevSM**2)

M1 = MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2)['M1']
M2 = MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2)['M2']
M3 = MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2)['M3']

vv12 = MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2)['V12']
vv22 = MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2)['V22']
vv32 = MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2)['V32']    
uu11 = MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2)['U11']
uu21 = MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2)['U21']
uu31 = MATRIXCHIDIAG(vevSM, MN, MPsi, Meta, l1, l2, y1, y2)['U31']

L1 = neutrino_analytic3.LAMBDA(M1,MS1,vv12,uu11)+neutrino_analytic3.LAMBDA(M2,MS1,vv22,uu21)\
    +neutrino_analytic3.LAMBDA(M3,MS1,vv32,uu31)

L2 = neutrino_analytic3.LAMBDA(M1,MS2,vv12,uu11)+neutrino_analytic3.LAMBDA(M2,MS2,vv22,uu21)\
    +neutrino_analytic3.LAMBDA(M3,MS2,vv32,uu31)

L3 = neutrino_analytic3.LAMBDA(M1,MS3,vv12,uu11)+neutrino_analytic3.LAMBDA(M2,MS3,vv22,uu21)\
    +neutrino_analytic3.LAMBDA(M3,MS3,vv32,uu31)

#####  ALGORITM: Particular case(see): inversion-neutrino-matrix.nb ######################
a11 = 0.0
a12 = 0.0 
### Random -> FREE PARAMETERS
a21 = np.exp(np.random.uniform(np.log(10**(-6)),np.log(10**(0))))
a22 = np.exp(np.random.uniform(np.log(10**(-6)),np.log(10**(0))))
a31 = np.exp(np.random.uniform(np.log(10**(-6)),np.log(10**(0))))    
a32 = np.exp(np.random.uniform(np.log(10**(-6)),np.log(10**(0))))
### Solved in Mathematica
b11 = -((a32*mnu2*U12 - a22*mnu3*U13)/(a22*a31*L1 - a21*a32*L1))   
b21 = -((a32*mnu2*U22 - a22*mnu3*U23)/(a22*a31*L1 - a21*a32*L1))
b31 = -((a32*mnu2*U32 - a22*mnu3*U33)/(a22*a31*L1 - a21*a32*L1))
b12 = (a31*mnu2*U12 - a21*mnu3*U13)/(a22*a31*L2 - a21*a32*L2)
b22 = (a31*mnu2*U22 - a21*mnu3*U23)/(a22*a31*L2 - a21*a32*L2)
b32 = (a31*mnu2*U32 - a21*mnu3*U33)/(a22*a31*L2 - a21*a32*L2)

xdict.blocks['YFR1IN'].entries[1]='%.6E    # YfR1(1)' %a11 
xdict.blocks['YFR1IN'].entries[2]='%.6E    # YfR1(2)' %a21
xdict.blocks['YFR1IN'].entries[3]='%.6E    # YfR1(3)' %a31
xdict.blocks['YFR2IN'].entries[1]='%.6E    # YfR2(1)' %a12 
xdict.blocks['YFR2IN'].entries[2]='%.6E    # YfR2(2)' %a22
xdict.blocks['YFR2IN'].entries[3]='%.6E    # YfR2(3)' %a32

xdict.blocks['YHR1IN'].entries[1]='%.6E    # YhR1(1)' %b11 
xdict.blocks['YHR1IN'].entries[2]='%.6E    # YhR1(2)' %b21
xdict.blocks['YHR1IN'].entries[3]='%.6E    # YhR1(3)' %b31
xdict.blocks['YHR2IN'].entries[1]='%.6E    # YhR2(1)' %b12 
xdict.blocks['YHR2IN'].entries[2]='%.6E    # YhR2(2)' %b22
xdict.blocks['YHR2IN'].entries[3]='%.6E    # YhR2(3)' %b32

nu1 = neutrino_analytic3.MATRIX_NU_DIAG(b11,b12,0,b21,b22,0,a11,a12,0,a21,a22,0,MS1,MS2,M1,M2,M3,vv12,vv22,uu11,uu21)
print("Rutine masses:")
print(nu1)

#Write the Leshouches file
pyslha.writeSLHAFile('LesHouches.in.T12B_low',xdict)    


#run SPheno
spheno = subprocess.getoutput('../.././SPheno-4.0.3/bin/SPhenoT12B LesHouches.in.T12B_low')    
SPheno_output = subprocess.getoutput('cat SPheno.spc.T12B')

##++++++++++ Neutrinos +++++++++++++++++++++++++++++++++
if SPheno_output.split('# Fk')[1].split()[1] == "NaN":
    mv1 = 0.
else: 
    mv1 = eval(SPheno_output.split('# Fk')[1].split()[1])           
mv2 = eval(SPheno_output.split('# Fk')[1].split()[5])       
mv3 = eval(SPheno_output.split('# Fk')[1].split()[9])  

print("SPHENO masses:")
print("mv1=",mv1,"mv2=",mv2,"mv3=",mv3)

x.append([L6,mv1,mv2,mv3])
    
x=np.asarray(x)
xd=pd.DataFrame(x,columns=['L6','mv1','mv2','mv3'])

t2=time.time() 
#print ("The program spent", t2-t1, "s running",Num,"times")          

Experimental masses:
mnu1= 1e-20 mnu2= 8.97561323708e-12 mnu3= 5.06809986249e-11
Rutine masses:
(0.0, 5.5511151231257827e-17, 5.6593766998435584e-09)
SPHENO masses:
mv1= 2.74927965e-16 mv2= 1.13987842e-12 mv3= 5.28202242e-12


In [72]:
xd

L6           mv1           mv2           mv3
0  0.013  4.073351e-16  8.877313e-13  6.347176e-12

In [38]:
SPheno_output.splitlines()

['# SUSY Les Houches Accord 2 - T12B Spectrum + Decays + Flavour Observables',
 '# SPheno module generated by SARAH',
 '# ----------------------------------------------------------------------',
 '# SPheno v4.0.3  ',
 '#   W. Porod, Comput. Phys. Commun. 153 (2003) 275-315, hep-ph/0301101',
 '#   W. Porod, F.Staub, Comput.Phys.Commun.183 (2012) 2458-2469, arXiv:1104.1573',
 '# SARAH: 4.14.4  ',
 '#   F. Staub; arXiv:0806.0538 (online manual)',
 '#   F. Staub; Comput. Phys. Commun. 181 (2010) 1077-1086; arXiv:0909.2863',
 '#   F. Staub; Comput. Phys. Commun. 182 (2011)  808-833; arXiv:1002.0840',
 '#   F. Staub; Comput. Phys. Commun. 184 (2013)  1792-1809; arXiv:1207.0906',
 '#   F. Staub; Comput. Phys. Commun. 185 (2014)  1773-1790; arXiv:1309.7223 ',
 '# Including the calculation of flavor observables based on FlavorKit ',
 '#   W. Porod, F. Staub, A. Vicente; Eur.Phys.J. C74 (2014) 8, 2992; arXiv:1405.1434 ',
 '# Two-loop mass corrections to Higgs fields based on ',
 '#   M. D. Goods

$$ m_{s_i}^2 = \mu_{s_i}^2 + \dfrac{\lambda_{S_iH}}{2}v^2$$